[jit] dispatch and expose linear op

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20284 make linear avaiable in tracing
* **#20039 dispatch and expose linear op**
* #19988 split canonicalize_ops, make a decompose pass
* #19987 lower batchmm to non-diff optimization

Summary:
This expose the linear directly when we called the functional interface, the ATen linear op already do the same thing with the functional interface, so there's really no need to duplicate the code. Also, this will expose the higher level `aten::linear` op up until the custom fusion, so that different backends could know the high level information

Test Plan:
Test the linear op is correctly decomposed in the decomposition pass, which also means it does not get decomposed before that pass. 


Currently being blocked from landing by https://github.com/pytorch/pytorch/issues/19769 and https://github.com/pytorch/pytorch/issues/20734. 

Differential Revision: [D15190354](https://our.internmc.facebook.com/intern/diff/D15190354)

wanchaol committed:
dispatch and expose linear op

wanchaol committed:
Update on "dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
replace inlineGraph on "dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
update callsite to linear on "dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
small update on test on "dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
fix on unspecialized input on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
add more test case on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

bwasti commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1454,6 +1454,11 @@ class ShapePropagator {
       }
+    } else if (node->matches("aten::linear(Tensor input, Tensor weight, Tensor? bias) -> Tensor")) {
+      if (auto type = input_type(0)) {
+        node->output()->setType(type);
```
is `type` here a complete tensor or just a dimensioned tensor?

the input shape is not the same as the output shape right?

wanchaol replied:
hmm you are right, actually it should be a complete tensor and the output shape should be different. seems like bilinear below is also wrong, will fix it.

bwasti reviewed (approved):
generally looks good, one concern re shape inference

wanchaol committed:
reorg code, add linear AD formula on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
update shape prop on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
update ad formula, add constructor to compilation unit on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

zdevito reviewed (commented):
I don't see any problems here, but because this shuffles around where decompositions happen, I am concerned about test coverage and performance. 

* Can we make sure the derivative formula for linear is actually being testing
* @apaszke can you take a look at this and see if anything jumps out as bugs? Context is that mkldnn backend has good fused linear performance and we are splitting the op up before it can be seen.

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1661,6 +1661,15 @@ class ShapePropagator {
+      SHAPE_ASSERT(weight_type->sizes().size() == 2 && sizes.size() >= 2);
+      sizes.at(last_dim) = weight_type->sizes()[0];
+      node->output()->setType(input_type->withSizes(sizes));
+      return true;
```
I don't think we need to add complete shape prop, since we're not really using it at this point. Also, this looks like the same case as for `aten::mm`, so maybe we could avoid duplicating them?

wanchaol replied:
`aten::mm` section requires the input to be 2, where `aten::linear` have 3.. I think I can also do partial shape prop like the `aten::bilinear` defined [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/shape_analysis.cpp#L1470), but this also requires additional code (which is a bit duplicated)..

apaszke replied:
Fair enough, let's leave it.

apaszke commented on torch/csrc/jit/passes/decompose_ops.cpp:
```diff
@@ -21,8 +27,39 @@ static bool DecomposeOps(Block* block) {
+      int ndim = input_type->dim();
+      Value* new_output = nullptr;
+      if (ndim == 2 && bias->type()->isSubtypeOf(TensorType::get())) {
+        // if ndim == 2 and bias is statically defined, dispatch to addmm decomposition
```
_statically defined_ is not very well defined here. You're only checking that we can refine its type, which doesn't tell you much.

wanchaol replied:
This is where I steal it from the [fuser code](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/graph_fuser.cpp#L156), where is tells you if it is statically defined (statically undefined needs more check though). because we are only check the input value types, I guess it would be sufficient to distinguish it. 

The next step after this PR, I will move the `batchnorm` and `layernorm` decomposition to here as well, then I can use the util function defined above.

apaszke replied:
My point is that the term _statically defined_ is not used correctly. I'm not even 100% sure what does that mean

wanchaol replied:
From my point of view, `statically defined` tensor argument means that the value passed in is neither a `None` constant nor a `Optional[Tensor]` type, which we could statically know the type passed in rather than a dynamically unknown optional type. So `isSubtypeOf(TensorType)` should guarantee this definition

apaszke reviewed (commented):
1. We should check the AD formulas
2. This likely breaks a lot of the fusions we used to do. We want to make sure that the bias additions in LSTM get fused, so the fuser needs to be taught to decompose `addmm`s.

wanchaol commented:
Thanks @zdevito @apaszke ! re concerns on ad formula and performance:

1. the ad formula is already been checked [here](https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L13461), it's checking the `input.dim() == 2` pathway, I will add one more test case to cover the matmul AD pathway
2. I believe by putting the decomposition pass before fuser and the PR added the decomposition from linear to `addmm`, it should preserve the old behavior as we did the bias add fusion, I checked the forward and backward graph for lstm, it remains the same, so it should be good on fusion and not regressed the LSTM performance. I will make sure to run enough benchmarks before landing this.

wanchaol committed:
address comments on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
fix tests on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
change profiled shapes since we directly call the cpp function on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
minor update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol commented:
I checked that the LSTM graph is the same before and after this PR, except the unique names of values are different:

https://gist.github.com/wanchaol/bcd988809a192c56f55319fc6a305637/revisions 

Performance: shows no noticeable difference between master and master + this PR: 

**On Master**:

<img width="1662" alt="before" src="https://user-images.githubusercontent.com/9443650/57363986-128ed000-7137-11e9-89bf-a242b3038676.png">

**Master + this stacked PR**:

<img width="1662" alt="after" src="https://user-images.githubusercontent.com/9443650/57363992-1589c080-7137-11e9-9c93-f1485eb0cbc5.png">

So I think this PR is not regressed our current fusion :)

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
small AD optimize on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

apaszke commented:
Ok I see you have made changes to how the ops are decomposed, so I'd need to read the prior patches to make sure this works ok. If you double checked that we're not loosing LSTM fusion + mm batching, then it might be good to go. I'll try to catch up with the whole stack soon.

wanchaol committed:
add onnx symbolic for linear on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
rebase with master on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

wanchaol committed:
Update on "[jit] dispatch and expose linear op"

dispatch and expose linear op

gh-metadata: pytorch pytorch 20039 gh/wanchaol/4/head

XiaobingSuper commented:
@wanchaol, Do you have any progress about this PR?

wanchaol commented:
@XiaobingSuper this is currently blocked from landing by issue #19769 and #20734, we don't have a good solution to those currently, so this might not be land until we could solve them.

XiaobingSuper commented:
@wanchaol , #19769 seems fixed, do you have any plan to do it?

wanchaol commented:
> @wanchaol , #19769 seems fixed, do you have any plan to do it?

@XiaobingSuper sorry the issue was closed to dedup, but the underlying problem is not fixed yet, which is https://github.com/pytorch/pytorch/issues/20734. 

But I think we recently figured out a plan on how to fix it and I will let you know about the progress once we start working on it :)
